presto-plan-checker-providers/src/main/java/com/facebook/presto/plancheckerproviders/nativechecker/NativePlanChecker.java

```diff
@@ -0,0 +1,182 @@
+    {
+        Set<Node> nodes = nodeManager.getAllNodes();
+        Node sidecarNode = nodes.stream().filter(Node::isCoordinatorSidecar).findFirst()
+                .orElseThrow(() -> new PrestoException(NOT_FOUND, "could not find native sidecar node"));
+        try {
+            // endpoint to retrieve session properties from native worker
+            return new URL(sidecarNode.getHttpUri().toString() + PLAN_CONVERSION_ENDPOINT);
+        }
+        catch (MalformedURLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void validate(PlanNode planNode, WarningCollector warningCollector)
+    {
+        // NO-OP, only validating fragments
+    }
+
+    @Override
+    public void validateFragment(SimplePlanFragment planFragment, WarningCollector warningCollector)
+    {
+        try {
+            if (!planFragment.getPartitioning().isCoordinatorOnly() && !isInternalSystemConnector(planFragment.getRoot())) {
+                runValidation(planFragment);
+            }
+            else if (LOG.isDebugEnabled()) {
+                LOG.debug("Skipping Native Plan Validation for plan fragment id: %s", planFragment.getId());
+            }
+        }
+        catch (final PrestoException e) {
+            if (config.isQueryFailOnError()) {
+                throw e;
+            }
+        }
+        catch (final Exception e) {
+            LOG.error(e, "Failed to run native plan validation");
+            if (config.isQueryFailOnError()) {
+                throw new IllegalStateException("Failed to run native plan validation", e);
+            }
+        }
+    }
+
+    private boolean isInternalSystemConnector(PlanNode planNode)
+    {
+        try {
+            return planNode.accept(new CheckInternalVisitor(), null);
+        }
+        catch (final IllegalArgumentException e) {
+            // An InternalPlanNode will not allow a `PlanVisitor`, allow it to pass
+            return false;
+        }
+    }
+
+    private void runValidation(SimplePlanFragment planFragment)
+            throws IOException
+    {
+        LOG.debug("Starting native plan validation for plan fragment id: %s", planFragment.getId());
+
+        String requestBodyJson = planFragmentJsonCodec.toJson(planFragment);
+        LOG.debug("Native validation for plan fragment: %s", requestBodyJson);
+
+        Request request = buildRequest(requestBodyJson);
+
+        try (Response response = httpClient.newCall(request).execute()) {
+            if (!response.isSuccessful()) {
+                String responseBody = response.body() != null ? response.body().string() : "{}";
+                LOG.error("Native plan checker failed with code: %d, response: %s", response.code(), responseBody);
+                if (config.isQueryFailOnError()) {
+                    throw new PrestoException(QUERY_REJECTED, "Query failed by native plan checker with code: " + response.code() + ", response: " + responseBody);
```

Comment:
So you mean the response from the side car should have an error code we use here instead of `QUERY_REJECTED`? The error message is propagated and shown to the user.

Reply:
The error that the user sees shouldn't be QUERY_REJECTED.  It should be whatever error caused the plan checking to fail.